File: backend/app/schemas/analytics.py
from pydantic import BaseModel


# ============================================================
# User Statistics
# ============================================================

class UserAnalyticsResponse(BaseModel):

    total_users: int

    active_users: int

    inactive_users: int

    total_students: int

    total_faculty: int

    total_institution_admins: int

    total_admins: int


# ============================================================
# Learning Statistics
# ============================================================

class LearningAnalyticsResponse(BaseModel):

    total_lessons: int

    completed_lessons: int

    lesson_completion_percentage: float


# ============================================================
# Quiz Statistics
# ============================================================

class QuizAnalyticsResponse(BaseModel):

    total_quizzes: int

    total_attempts: int

    completed_attempts: int

    passed_attempts: int

    average_percentage: float

    pass_percentage: float


# ============================================================
# Simulation Statistics
# ============================================================

class SimulationAnalyticsResponse(BaseModel):

    total_simulations: int

    total_responses: int

    completed_simulations: int

    average_percentage: float


# ============================================================
# Disaster Analytics
# ============================================================

class DisasterAnalyticsResponse(BaseModel):

    disaster_id: int

    disaster_name: str

    total_lessons: int

    completed_lessons: int

    lesson_completion_percentage: float

    total_quizzes: int

    quiz_attempts: int

    quiz_average_percentage: float

    total_simulations: int

    simulation_responses: int


# ============================================================
# Overall Analytics
# ============================================================

class AdminAnalyticsResponse(BaseModel):

    users: UserAnalyticsResponse

    learning: LearningAnalyticsResponse

    quizzes: QuizAnalyticsResponse

    simulations: SimulationAnalyticsResponse


# ============================================================
# Analytics Dashboard
# ============================================================

class AnalyticsDashboardResponse(BaseModel):

    users: UserAnalyticsResponse

    learning: LearningAnalyticsResponse

    quizzes: QuizAnalyticsResponse

    simulations: SimulationAnalyticsResponse

    disasters: list[DisasterAnalyticsResponse]

File: frontend/src/api/analytics.ts
import apiClient from "./client";

export interface UserAnalytics {
    total_users: number;
    active_users: number;
    inactive_users: number;
    total_students: number;
    total_faculty: number;
    total_institution_admins: number;
    total_admins: number;
}

export interface LearningAnalytics {
    total_lessons: number;
    completed_lessons: number;
    lesson_completion_percentage: number;
}

export interface QuizAnalytics {
    total_quizzes: number;
    total_attempts: number;
    completed_attempts: number;
    passed_attempts: number;
    average_percentage: number;
    pass_percentage: number;
}

export interface SimulationAnalytics {
    total_simulations: number;
    total_responses: number;
    completed_simulations: number;
    average_percentage: number;
}

export interface DisasterAnalytics {
    disaster_id: number;
    disaster_name: string;
    total_lessons: number;
    completed_lessons: number;
    lesson_completion_percentage: number;
    total_quizzes: number;
    quiz_attempts: number;
    quiz_average_percentage: number;
    total_simulations: number;
    simulation_responses: number;
}

export interface AnalyticsDashboard {
    users: UserAnalytics;
    learning: LearningAnalytics;
    quizzes: QuizAnalytics;
    simulations: SimulationAnalytics;
    disasters: DisasterAnalytics[];
}


export const getAnalyticsDashboard =
    async (): Promise<AnalyticsDashboard> => {

        const response =
            await apiClient.get<AnalyticsDashboard>(
                "/analytics/dashboard"
            );

        return response.data;
    };